tower-ready-cache/src/lib.rs

```diff
@@ -0,0 +1,276 @@
+use futures::{stream, Async, Future, Poll, Stream};
+pub use indexmap::Equivalent;
+use indexmap::IndexMap;
+use log::{debug, trace};
+use std::hash::Hash;
+use tokio_sync::oneshot;
+use tower_service::Service;
+use tower_util::Ready;
+
+/// Caches services, segregating ready and unready services.
+#[derive(Debug)]
+pub struct ReadyCache<K, S, Req>
+where
+    K: Eq + Hash,
+{
+    ready_services: IndexMap<K, S>,
+    unready_services: stream::FuturesUnordered<UnreadyService<K, S, Req>>,
+    cancelations: IndexMap<K, oneshot::Sender<()>>,
+}
+
+pub type Error = Box<dyn std::error::Error + Send + Sync>;
+
+enum UnreadyError {
+    Inner(Error),
+    Canceled,
+}
+
+/// A Future that becomes satisfied when an `S`-typed service is ready.
+///
+/// May fail due to cancelation, i.e. if the service is removed from discovery.
+#[derive(Debug)]
+struct UnreadyService<K, S, Req> {
+    key: Option<K>,
+    cancel: oneshot::Receiver<()>,
+    ready: tower_util::Ready<S, Req>,
+}
+
+// === ReadyCache ===
+
+impl<K, S, Req> Default for ReadyCache<K, S, Req>
+where
+    K: Eq + Hash,
+    S: Service<Req>,
+    S::Error: Into<Error>,
+{
+    fn default() -> Self {
+        Self {
+            ready_services: IndexMap::default(),
+            cancelations: IndexMap::default(),
+            unready_services: stream::FuturesUnordered::new(),
+        }
+    }
+}
+
+impl<K, S, Req> ReadyCache<K, S, Req>
+where
+    K: Eq + Hash,
+{
+    /// Returns the number of services in the ready set.
+    pub fn ready_len(&self) -> usize {
+        self.ready_services.len()
+    }
+
+    /// Returns the number of services in the unready set.
+    pub fn unready_len(&self) -> usize {
+        self.unready_services.len()
+    }
+
+    /// Returns true iff the given key is in the unready set.
+    pub fn unready_contains<Q: Hash + Equivalent<K>>(&self, key: &Q) -> bool {
+        self.cancelations.contains_key(key)
+    }
+
+    /// Attempts to cancel a service in the unready set.
+    ///
+    /// The service will be removed from the cache when `poll_unready` is called.
+    ///
+    /// Returns true iff the referenced service was unready and it was
+    /// succsesfully canceled.
+    pub fn cancel_unready<Q: Hash + Equivalent<K>>(&mut self, key: &Q) -> bool {
```

Comment:
It doesn't look like this method is used externally anywhere (rather, everyone calls `evict`). Are you expecting this to ever be called specifically?